src/databricks/labs/ucx/azure/access.py

```diff
@@ -267,27 +307,57 @@ def create_uber_principal(self, prompts: Prompts):
-                storage_account_info.append(storage)
+                storage_accounts.append(storage)
         logger.info("Creating service principal")
-        uber_principal = self._azurerm.create_service_principal(uber_principal_name)
-        self._create_scope(uber_principal, inventory_database)
-        config.uber_spn_id = uber_principal.client.client_id
-        logger.info(
-            f"Created service principal of client_id {config.uber_spn_id}. " f"Applying permission on storage accounts"
-        )
         try:
+            uber_principal = self._azurerm.create_service_principal(uber_principal_name)
             self._apply_storage_permission(
-                uber_principal.client.object_id, "STORAGE_BLOB_DATA_CONTRIBUTOR", *storage_account_info
+                uber_principal.client.object_id, "STORAGE_BLOB_DATA_CONTRIBUTOR", *storage_accounts
             )
-            self._installation.save(config)
-            self._update_cluster_policy_with_spn(policy_id, storage_account_info, uber_principal, inventory_database)
-            self._update_sql_dac_with_spn(storage_account_info, uber_principal, inventory_database)
-        except PermissionError:
-            self._azurerm.delete_service_principal(uber_principal.client.object_id)
-        logger.info(f"Update UCX cluster policy {policy_id} with spn connection details for storage accounts")
+            secret = self._create_and_get_secret_for_uber_principal(uber_principal, inventory_database)
+            secret_identifier = f"secrets/{inventory_database}/{secret.key}"
+            self._update_cluster_policy_with_spn(policy_id, uber_principal, secret_identifier, storage_accounts)
+            self._update_sql_dac_with_spn(uber_principal.client.client_id, secret_identifier, storage_accounts)
+        except (PermissionError, NotFound):
+            logger.error("Failed to create service principal", exc_info=True)
+            self._delete_uber_principal()  # Clean up dangling resources
+            return
+        config.uber_spn_id = uber_principal.client.client_id
+        self._installation.save(config)
+        logger.info(f"Created service principal ({config.uber_spn_id}) with access to used storage accounts.")
+        logger.info(f"Updated UCX cluster policy {policy_id} with spn connection details for storage accounts")
+
+    def _delete_uber_principal(self):
+        config = self._installation.load(WorkspaceConfig)
+        if config.uber_spn_id is not None:
+            used_storage_accounts = self._get_storage_accounts()
+            storage_accounts = []
+            for storage in self._azurerm.storage_accounts():
+                if storage.name in used_storage_accounts:
+                    storage_accounts.append(storage)
+            for storage_account in storage_accounts:
+                try:
+                    self._azurerm.delete_storage_permission(str(storage_account.id), principal_id=config.uber_spn_id)
+                except NotFound:
```

Comment:
These are databricks exceptions while azure resources are deleted

Reply:
because we use Databricks SDK API client. that's fine